File: app/src/main/java/com/hjc/reader/bean/db/HistoryDataBase.java
package com.hjc.reader.bean.db;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

@Database(entities = {History.class}, version = 1, exportSchema = false)
public abstract class HistoryDataBase extends RoomDatabase {

    private static final String DB_NAME = "search_history";

    private static HistoryDataBase historyDataBase;

    public static HistoryDataBase getInstance(Context context) {
        if (historyDataBase == null) {
            synchronized (HistoryDataBase.class) {
                historyDataBase = Room.databaseBuilder(
                        context.getApplicationContext(),
                        HistoryDataBase.class,
                        DB_NAME
                ).allowMainThreadQueries().build();
            }
        }
        return historyDataBase;
    }

    public abstract HistoryDao getHistoryDao();
}

File: module-main/src/main/java/com/hjc/learn/main/entity/HistoryDataBase.kt
package com.hjc.learn.main.entity

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase

@Database(entities = [History::class], version = 1)
abstract class HistoryDataBase : RoomDatabase() {

    // 必须包含一个具有0个参数且返回带@Dao注释的类的抽象方法
    abstract fun getHistoryDao(): HistoryDao

    companion object {
        @Volatile
        private var instance: HistoryDataBase? = null

        fun getInstance(context: Context): HistoryDataBase = instance ?: synchronized(this) {
            instance ?: buildDatabase(context).also { instance = it }
        }

        private fun buildDatabase(context: Context) =
            Room.databaseBuilder(
                context.applicationContext,
                HistoryDataBase::class.java, "search_history.db"
            )
                .allowMainThreadQueries()
                .build()
    }
}